File: backend/stores/material_store.py
"""服务器端素材库 — 替代浏览器 MaterialStore

会话级素材存储，每条素材绑定 source_index 用于溯源。
支持自动过期（默认 30 分钟无活动）。
"""
import uuid
import time
import threading
from typing import Optional
from dataclasses import dataclass, field

from backend.config import SESSION_TTL


@dataclass
class SourceRef:
    """素材溯源引用"""
    id: str
    file: str
    section: str
    timestamp: float = field(default_factory=time.time)
    text_snippet: str = ""


@dataclass
class MaterialSession:
    """单个会话的素材库"""
    session_id: str
    created_at: float = field(default_factory=time.time)
    last_active: float = field(default_factory=time.time)

    # 身份信息
    identity: dict = field(default_factory=lambda: {
        "name": None, "phone": None, "email": None, "city": None,
        "target_job": None, "salary": None, "onboard_time": None,
        "birth_date": None, "age": None, "gender": None,
    })

    # 教育经历
    education: list = field(default_factory=list)

    # 工作经历
    work_experience: list = field(default_factory=list)

    # 项目经历
    projects: list = field(default_factory=list)

    # 技能
    skills: list = field(default_factory=list)

    # 证书
    certificates: list = field(default_factory=list)

    # 语言能力
    languages: list = field(default_factory=list)

    # 自我评价
    self_assessment: Optional[str] = None

    # JD 数据
    jd: dict = field(default_factory=lambda: {
        "raw_text": None,
        "keywords": {"hard_skills": [], "soft_skills": [], "industry": []},
        "requirements": [],
        "responsibilities": [],
    })

    # 问卷回答
    qa_entries: list = field(default_factory=list)

    # 原始文件元数据
    source_files: list = field(default_factory=list)

    # 溯源索引
    _source_index: int = 0
    _locked: bool = False

    def next_ref(self, file_name: str, section: str, snippet: str = "") -> SourceRef:
        """生成新的溯源引用"""
        self._source_index += 1
        return SourceRef(
            id=f"src-{self._source_index}",
            file=file_name,
            section=section,
            text_snippet=snippet[:200],
        )

    def set_identity(self, field: str, value: str, file_name: str = "unknown") -> bool:
        if self._locked:
            return False
        ref = self.next_ref(file_name, f"个人信息/{field}", value)
        self.identity[field] = {"value": str(value).strip(), "sources": [ref]}
        return True

    def get_identity(self, field: str) -> Optional[str]:
        v = self.identity.get(field)
        return v["value"] if v else None

    def add_education(self, entry: dict, file_name: str = "unknown") -> bool:
        if self._locked:
            return False
        ref = self.next_ref(file_name, "教育经历", str(entry.get("school", "")))
        self.education.append({**entry, "source": ref})
        return True

    def add_work_experience(self, entry: dict, file_name: str = "unknown") -> bool:
        if self._locked:
            return False
        ref = self.next_ref(file_name, "工作经历", str(entry.get("company", "")))
        self.work_experience.append({**entry, "source": ref})
        return True

    def add_project(self, entry: dict, file_name: str = "unknown") -> bool:
        if self._locked:
            return False
        ref = self.next_ref(file_name, "项目经历", str(entry.get("name", "")))
        self.projects.append({**entry, "source": ref})
        return True

    def add_skill(self, entry: dict, file_name: str = "unknown") -> bool:
        if self._locked:
            return False
        ref = self.next_ref(file_name, "技能", str(entry.get("name", "")))
        self.skills.append({**entry, "source": ref})
        return True

    def add_certificate(self, entry: dict, file_name: str = "unknown") -> bool:
        if self._locked:
            return False
        ref = self.next_ref(file_name, "证书", str(entry.get("name", "")))
        self.certificates.append({**entry, "source": ref})
        return True

    def set_self_assessment(self, text: str, file_name: str = "unknown") -> bool:
        if self._locked:
            return False
        ref = self.next_ref(file_name, "自我评价", text)
        self.self_assessment = {"value": text, "source": ref}
        return True

    def set_jd(self, jd_data: dict) -> None:
        self.jd.update(jd_data)

    def load_from_extraction(self, extracted_data: dict, file_name: str = "unknown") -> None:
        """从提取结果批量导入素材库"""
        if self._locked:
            return

        basic = extracted_data.get("basic_info", {})
        for field in ["name", "phone", "email", "city", "target_job", "expect_salary", "onboard_time"]:
            if basic.get(field):
                self.set_identity(field, basic[field], file_name)

        for edu in extracted_data.get("education", []):
            self.add_education(edu, file_name)

        for work in extracted_data.get("work_experience", []):
            self.add_work_experience(work, file_name)

        for proj in extracted_data.get("projects", []):
            self.add_project(proj, file_name)

        for skill in extracted_data.get("skills", []):
            self.add_skill(skill, file_name)

        for cert in extracted_data.get("certificates", []):
            self.add_certificate(cert, file_name)

        sa = extracted_data.get("self_assessment")
        if sa:
            self.set_self_assessment(sa, file_name)

        file_meta = {
            "name": file_name,
            "type": extracted_data.get("file_type", "unknown"),
            "method": extracted_data.get("method", "unknown"),
            "extracted_at": time.time(),
        }
        self.source_files.append(file_meta)

    def lock(self) -> None:
        """锁定素材库，禁止修改"""
        self._locked = True

    def is_locked(self) -> bool:
        return self._locked

    def get_ai_context(self) -> str:
        """组装完整的 AI 上下文（用于 LLM prompt 注入）"""
        parts = []

        # 基础信息
        id_parts = []
        for field, key in [("name", "姓名"), ("phone", "电话"), ("email", "邮箱"),
                           ("city", "城市"), ("target_job", "目标岗位"), ("salary", "期望薪资")]:
            v = self.identity.get(field)
            if v and v.get("value"):
                id_parts.append(f"{key}: {v['value']}")
        if id_parts:
            parts.append("【基础信息】\n" + "\n".join(id_parts))

        # 教育
        if self.education:
            edu_lines = []
            for e in self.education:
                parts_list = []
                if e.get("school"): parts_list.append(e["school"])
                if e.get("major"): parts_list.append(e["major"])
                if e.get("degree"): parts_list.append(e["degree"])
                if e.get("start_date") or e.get("end_date"):
                    parts_list.append(f"{e.get('start_date','')} - {e.get('end_date','')}")
                edu_lines.append(" | ".join(parts_list))
            parts.append("【教育经历】\n" + "\n".join(edu_lines))

        # 工作
        if self.work_experience:
            work_lines = []
            for w in self.work_experience:
                wl = []
                if w.get("company"): wl.append(w["company"])
                if w.get("position"): wl.append(w["position"])
                if w.get("start_date") or w.get("end_date"):
                    wl.append(f"{w.get('start_date','')} - {w.get('end_date','')}")
                work_lines.append(" | ".join(wl))
                if w.get("duties"):
                    work_lines.append(f"  职责: {w['duties']}")
                if w.get("achievements"):
                    for a in w["achievements"]:
                        work_lines.append(f"  成果: {a}")
            parts.append("【工作经历】\n" + "\n".join(work_lines))

        # 项目
        if self.projects:
            proj_lines = []
            for p in self.projects:
                pl = []
                if p.get("name"): pl.append(p["name"])
                if p.get("role"): pl.append(f"角色: {p['role']}")
                proj_lines.append(" | ".join(pl))
                if p.get("description"):
                    proj_lines.append(f"  描述: {p['description']}")
                if p.get("results"):
                    proj_lines.append(f"  成果: {p['results']}")
            parts.append("【项目经历】\n" + "\n".join(proj_lines))

        # 技能
        if self.skills:
            skill_names = [s.get("name", "") for s in self.skills if s.get("name")]
            parts.append(f"【技能】{', '.join(skill_names)}")

        # 证书
        if self.certificates:
            cert_names = [c.get("name", "") for c in self.certificates if c.get("name")]
            parts.append(f"【证书】{', '.join(cert_names)}")

        # 自我评价
        if self.self_assessment and self.self_assessment.get("value"):
            parts.append(f"【自我评价】\n{self.self_assessment['value']}")

        # JD
        if self.jd.get("raw_text"):
            parts.append(f"【JD原文】\n{self.jd['raw_text']}")

        return "\n\n".join(parts)


class MaterialStore:
    """会话级素材库管理器（单例）"""

    _instance: Optional["MaterialStore"] = None

    def __init__(self):
        self._sessions: dict[str, MaterialSession] = {}
        self._lock = threading.Lock()
        self._cleanup_interval = 300  # 5分钟清理一次

    @classmethod
    def get_instance(cls) -> "MaterialStore":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def create_session(self) -> MaterialSession:
        """创建新会话"""
        sid = str(uuid.uuid4())[:12]
        session = MaterialSession(session_id=sid)
        with self._lock:
            self._sessions[sid] = session
        return session

    def get_session(self, session_id: str) -> Optional[MaterialSession]:
        """获取会话（自动续期）"""
        session = self._sessions.get(session_id)
        if session is None:
            return None
        if time.time() - session.last_active > SESSION_TTL:
            self._sessions.pop(session_id, None)
            return None
        session.last_active = time.time()
        return session

    def delete_session(self, session_id: str) -> bool:
        with self._lock:
            return self._sessions.pop(session_id, None) is not None

    def cleanup_expired(self) -> int:
        """清理过期会话，返回清理数量"""
        now = time.time()
        expired = [
            sid for sid, s in self._sessions.items()
            if now - s.last_active > SESSION_TTL
        ]
        with self._lock:
            for sid in expired:
                self._sessions.pop(sid, None)
        return len(expired)

    @property
    def session_count(self) -> int:
        return len(self._sessions)


# 便捷模块级函数
def get_store() -> MaterialStore:
    return MaterialStore.get_instance()


def create_session() -> MaterialSession:
    return get_store().create_session()

File: js/material-store.js
/* ============================================================
   material-store.js — 用户原始素材库（只读结构化存储）
   所有AI输出必须绑定此库中的原文索引，禁止编造
   ============================================================ */

const MaterialStore = (() => {
  // Internal structured storage
  let store = {
    // Personal identity info
    identity: {
      name: null,           // { value, sources: [] }
      phone: null,
      email: null,
      city: null,
      targetJob: null,
      salary: null,
    },
    // Education entries
    education: [],          // [{ school, major, degree, period, raw, source }]
    // Work experience entries
    workExperience: [],     // [{ company, position, duties, period, raw, source }]
    // Project entries
    projects: [],           // [{ name, role, description, results, period, raw, source }]
    // Skills
    skills: [],             // [{ name, category, source }]
    // Certificates
    certificates: [],       // [{ name, source }]
    // Self-assessment
    selfIntro: null,        // { value, source }
    // JD structured data
    jd: {
      rawText: null,
      keywords: { hardSkills: [], softSkills: [], industry: [] },
      requirements: [],     // [{ type, value, source }]
      responsibilities: [], // [{ value, source }]
    },
    // Questionnaire answers (增信问卷，作为合法原始数据)
    qaEntries: [],          // [{ question, answer, source: 'questionnaire' }]
    // All raw source files metadata
    sourceFiles: [],        // [{ name, type, extractedAt, text }]
    // Source index for cross-referencing
    _sourceIndex: 0,
  };

  // Lock flag — once material is extracted, AI cannot modify it
  let materialLocked = false;

  // === Source reference generator ===
  function nextSourceRef(fileName, section) {
    store._sourceIndex++;
    return {
      id: `src-${store._sourceIndex}`,
      file: fileName || '用户输入',
      section: section || '通用',
      timestamp: Date.now(),
    };
  }

  // === Identity ===
  function setIdentity(field, value, fileName) {
    if (materialLocked) return false;
    const ref = nextSourceRef(fileName, `个人信息/${field}`);
    store.identity[field] = { value: String(value).trim(), sources: [ref] };
    return true;
  }

  function getIdentity(field) {
    return store.identity[field]?.value || null;
  }

  // === Education ===
  function addEducation(entry, fileName) {
    if (materialLocked) return false;
    const ref = nextSourceRef(fileName, '教育经历');
    store.education.push({ ...entry, source: ref, raw: entry.raw || entry.school || '' });
    return true;
  }

  function getEducation() { return [...store.education]; }

  // === Work Experience ===
  function addWorkExperience(entry, fileName) {
    if (materialLocked) return false;
    const ref = nextSourceRef(fileName, '工作经历');
    store.workExperience.push({ ...entry, source: ref, raw: entry.raw || entry.company || '' });
    return true;
  }

  function getWorkExperience() { return [...store.workExperience]; }

  // === Projects ===
  function addProject(entry, fileName) {
    if (materialLocked) return false;
    const ref = nextSourceRef(fileName, '项目经历');
    store.projects.push({ ...entry, source: ref, raw: entry.raw || entry.name || '' });
    return true;
  }

  function getProjects() { return [...store.projects]; }

  // === Skills ===
  function addSkill(name, category, fileName) {
    if (materialLocked) return false;
    const ref = nextSourceRef(fileName, '技能证书');
    store.skills.push({ name: String(name).trim(), category: category || '通用', source: ref });
    return true;
  }

  function getSkills() { return [...store.skills]; }
  function getSkillNames() { return store.skills.map(s => s.name); }

  // === Certificates ===
  function addCertificate(name, fileName) {
    if (materialLocked) return false;
    const ref = nextSourceRef(fileName, '证书');
    store.certificates.push({ name: String(name).trim(), source: ref });
    return true;
  }

  function getCertificates() { return [...store.certificates]; }
  function getCertificateNames() { return store.certificates.map(c => c.name); }

  // === Self Intro ===
  function setSelfIntro(value, fileName) {
    if (materialLocked) return false;
    const ref = nextSourceRef(fileName, '自我评价');
    store.selfIntro = { value: String(value).trim(), source: ref };
    return true;
  }

  function getSelfIntro() { return store.selfIntro?.value || null; }

  // === JD ===
  function setJDRawText(text, fileName) {
    if (materialLocked) return false;
    store.jd.rawText = String(text).trim();
    store.jd._source = nextSourceRef(fileName, 'JD原文');
  }

  function setJDKeywords(keywords, fileName) {
    if (materialLocked) return false;
    store.jd.keywords = {
      hardSkills: keywords.hardSkills || [],
      softSkills: keywords.softSkills || [],
      industry: keywords.industry || [],
    };
    store.jd._kwSource = nextSourceRef(fileName, 'JD关键词');
  }

  function addJDRequirement(type, value, fileName) {
    if (materialLocked) return false;
    const ref = nextSourceRef(fileName, 'JD要求');
    store.jd.responsibilities.push({ type, value, source: ref });
  }

  function getJD() { return { ...store.jd }; }
  function getJDRawText() { return store.jd.rawText || ''; }
  function getJDKeywords() { return { ...store.jd.keywords }; }

  // === QA Entries (增信问卷 — 合法原始数据) ===
  function addQAEntry(question, answer) {
    const ref = nextSourceRef('增信问卷', question);
    store.qaEntries.push({ question, answer: String(answer).trim(), source: ref });
    return true;
  }

  function getQAEntries() { return [...store.qaEntries]; }

  // === Source Files ===
  function addSourceFile(name, type, text) {
    store.sourceFiles.push({
      name,
      type,
      extractedAt: Date.now(),
      text: String(text || '').trim(),
    });
  }

  function getSourceFiles() { return [...store.sourceFiles]; }

  // === Bulk Import from Parsed Data ===
  function importParsedData(parsed, fileName) {
    if (materialLocked) return { success: false, reason: '素材库已锁定' };

    const imported = { identity: 0, education: 0, work: 0, projects: 0, skills: 0 };

    if (parsed.name) { setIdentity('name', parsed.name, fileName); imported.identity++; }
    if (parsed.phone) { setIdentity('phone', parsed.phone, fileName); imported.identity++; }
    if (parsed.email) { setIdentity('email', parsed.email, fileName); imported.identity++; }
    if (parsed.city) { setIdentity('city', parsed.city, fileName); imported.identity++; }
    if (parsed.targetJob) { setIdentity('targetJob', parsed.targetJob, fileName); imported.identity++; }
    if (parsed.salary) { setIdentity('salary', parsed.salary, fileName); imported.identity++; }

    if (parsed.education && Array.isArray(parsed.education)) {
      parsed.education.forEach(e => { addEducation(e, fileName); imported.education++; });
    }
    if (parsed.workExperience && Array.isArray(parsed.workExperience)) {
      parsed.workExperience.forEach(w => { addWorkExperience(w, fileName); imported.work++; });
    }
    if (parsed.projects && Array.isArray(parsed.projects)) {
      parsed.projects.forEach(p => { addProject(p, fileName); imported.projects++; });
    }
    if (parsed.skills && Array.isArray(parsed.skills)) {
      parsed.skills.forEach(s => addSkill(typeof s === 'string' ? s : s.name, typeof s === 'string' ? '通用' : s.category, fileName));
      imported.skills += parsed.skills.length;
    }
    if (parsed.certificates && Array.isArray(parsed.certificates)) {
      parsed.certificates.forEach(c => addCertificate(typeof c === 'string' ? c : c.name, fileName));
    }
    if (parsed.selfIntro) { setSelfIntro(parsed.selfIntro, fileName); }

    // Add to source files record
    if (parsed.rawText) {
      addSourceFile(fileName, 'document', parsed.rawText);
    }

    return { success: true, imported };
  }

  // === Import JD parsed data ===
  function importJDParsedData(parsed, fileName) {
    if (materialLocked) return { success: false, reason: '素材库已锁定' };
    if (parsed.rawText) setJDRawText(parsed.rawText, fileName);
    if (parsed.keywords) setJDKeywords(parsed.keywords, fileName);
    if (parsed.requirements) {
      parsed.requirements.forEach(r => addJDRequirement(r.type || '要求', r.value, fileName));
    }
    return { success: true };
  }

  // === Lock / Unlock ===
  function lock() { materialLocked = true; }
  function unlock() { materialLocked = false; }
  function isLocked() { return materialLocked; }

  // === Get all material as flat text for AI context ===
  function getAIContext() {
    const parts = [];

    const id = store.identity;
    const idParts = [];
    if (id.name?.value) idParts.push(`姓名: ${id.name.value}`);
    if (id.phone?.value) idParts.push(`电话: ${id.phone.value}`);
    if (id.email?.value) idParts.push(`邮箱: ${id.email.value}`);
    if (id.city?.value) idParts.push(`城市: ${id.city.value}`);
    if (id.targetJob?.value) idParts.push(`目标岗位: ${id.targetJob.value}`);
    if (id.salary?.value) idParts.push(`期望薪资: ${id.salary.value}`);
    if (idParts.length) parts.push('【个人信息】\n' + idParts.join('\n'));

    if (store.education.length) {
      parts.push('【教育经历】\n' + store.education.map(e =>
        `${e.school || ''} / ${e.major || ''} / ${e.degree || ''} / ${e.period || ''} [来源: ${e.source?.file || '未知'}]`
      ).join('\n'));
    }

    if (store.workExperience.length) {
      parts.push('【工作经历】\n' + store.workExperience.map(w =>
        `${w.company || ''} / ${w.position || ''} / ${w.period || ''}\n  职责: ${w.duties || w.raw || ''} [来源: ${w.source?.file || '未知'}]`
      ).join('\n'));
    }

    if (store.projects.length) {
      parts.push('【项目经历】\n' + store.projects.map(p =>
        `${p.name || ''} / ${p.role || ''} / ${p.period || ''}\n  描述: ${p.description || p.raw || ''} [来源: ${p.source?.file || '未知'}]`
      ).join('\n'));
    }

    if (store.skills.length) {
      parts.push('【技能证书】\n' + store.skills.map(s => s.name).join('、'));
    }

    if (store.selfIntro?.value) {
      parts.push(`【自我评价】\n${store.selfIntro.value}`);
    }

    if (store.jd.rawText) {
      parts.push('【JD原文】\n' + store.jd.rawText);
    }

    if (store.qaEntries.length) {
      parts.push('【增信问卷】\n' + store.qaEntries.map(q =>
        `Q: ${q.question}\nA: ${q.answer}`
      ).join('\n'));
    }

    return parts.join('\n\n---\n\n');
  }

  // === Get all source references ===
  function getAllSources() {
    const sources = [];
    const addSource = (s) => { if (s && !sources.find(x => x.id === s.id)) sources.push(s); };

    Object.values(store.identity).forEach(v => { if (v?.sources) v.sources.forEach(addSource); });
    store.education.forEach(e => addSource(e.source));
    store.workExperience.forEach(w => addSource(w.source));
    store.projects.forEach(p => addSource(p.source));
    store.skills.forEach(s => addSource(s.source));
    store.certificates.forEach(c => addSource(c.source));
    if (store.selfIntro?.source) addSource(store.selfIntro.source);
    if (store.jd._source) addSource(store.jd._source);
    store.qaEntries.forEach(q => addSource(q.source));

    return sources;
  }

  // === Reset ===
  function reset() {
    store = {
      identity: { name: null, phone: null, email: null, city: null, targetJob: null, salary: null },
      education: [], workExperience: [], projects: [], skills: [], certificates: [],
      selfIntro: null,
      jd: { rawText: null, keywords: { hardSkills: [], softSkills: [], industry: [] }, requirements: [], responsibilities: [] },
      qaEntries: [], sourceFiles: [], _sourceIndex: 0,
    };
    materialLocked = false;
  }

  // === Export for debugging ===
  function dump() { return JSON.parse(JSON.stringify(store)); }

  // === 从提取结果批量导入 ===
  function loadFromExtraction(extractedData, fileName) {
    if (materialLocked) return false;
    return importParsedData(extractedData, fileName);
  }

  // === 获取素材锁定状态 ===
  function isMaterialLocked() { return materialLocked; }

  // === 获取溯源映射（字段 → 原文行） ===
  function getSourceMap() {
    const map = {};
    const addEntry = (key, entry) => {
      if (!entry) return;
      const src = entry.source || (entry.sources && entry.sources[0]);
      if (src) {
        if (!map[key]) map[key] = [];
        map[key].push({ id: src.id, file: src.file, section: src.section, snippet: src.text_snippet || '' });
      }
    };

    addEntry('name', store.identity.name);
    addEntry('phone', store.identity.phone);
    addEntry('email', store.identity.email);
    addEntry('city', store.identity.city);
    addEntry('target_job', store.identity.targetJob);

    store.education.forEach((e, i) => addEntry(`education_${i}`, e));
    store.workExperience.forEach((w, i) => addEntry(`work_${i}`, w));
    store.projects.forEach((p, i) => addEntry(`project_${i}`, p));
    store.skills.forEach((s, i) => addEntry(`skill_${i}`, s));
    store.certificates.forEach((c, i) => addEntry(`cert_${i}`, c));

    return map;
  }

  return {
    // Identity
    setIdentity, getIdentity,
    // Education
    addEducation, getEducation,
    // Work
    addWorkExperience, getWorkExperience,
    // Projects
    addProject, getProjects,
    // Skills
    addSkill, getSkills, getSkillNames,
    // Certificates
    addCertificate, getCertificates, getCertificateNames,
    // Self Intro
    setSelfIntro, getSelfIntro,
    // JD
    setJDRawText, setJDKeywords, addJDRequirement, getJD, getJDRawText, getJDKeywords,
    // QA
    addQAEntry, getQAEntries,
    // Source files
    addSourceFile, getSourceFiles,
    // Bulk import
    importParsedData, importJDParsedData, loadFromExtraction,
    // Lock
    lock, unlock, isLocked, isMaterialLocked,
    // Context & Sources
    getAIContext, getAllSources, getSourceMap,
    // Lifecycle
    reset, dump,
  };
})();

window.MaterialStore = MaterialStore;
